test/suite/hook_rollback.py

```diff
@@ -0,0 +1,135 @@
+# IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+# [TEST_TAGS]
+# ignored_file
+# [END_TAGS]
+
+# hook_rollback.py
+#   Intentionally have random failures of rollback some fraction of the time:
+#     python run.py --hook rollback=FAILRATE base01
+#
+# FAILRATE should be a floating number less than one, for example, 0.25 to fail a quarter of the time.
+#
+# This hook is used to test a feature of the test suite, where tests with rollback errors are automatically
+# restarted.
+
+from __future__ import print_function
+
+import os, sys, wthooks, random, wiredtiger
+from wttest import WiredTigerTestCase
+
+# Print to /dev/tty for debugging, since anything extraneous to stdout/stderr will
+# cause a test error.
+def tty(s):
+    WiredTigerTestCase.tty(s)
+
+# These are the hook functions that are run when particular APIs are called.
+
+def session_begin_transaction_notify(ret, session, *args):
+    session.in_transaction = True
+    #tty('==> in txn, session={}'.format(session))
+
+def session_end_transaction_notify(ret, session, *args):
+    session.in_transaction = False
+    #tty('==> out txn, session={}'.format(session))
+
+# Here we patch a deficiency of our SWIG setup.  When a cursor is created via
+# a session.open_cursor() call, a unique Cursor python object is created.
+# However, that particular 'cursor' object doesn't know that 'session' created it,
+# only that there is a property 'session' that maps to the pointer WT_CURSOR->session
+# pointer.  The upshot is that cursor.session will not give us the same Python object
+# that created the cursor, though it will be a wrapper around the same C WT_SESSION.
+#
+# To fix this, we change the Cursor.session property, it was originally managed by SWIG,
+# now it just returns a session field, that we set immediately after the open_cursor.
+# In an ideal world, we'd change our SWIG configuration to do the right thing for
+# Cursor.session and Session.connection, but I don't think any of the regular tests care about it.
+
+wiredtiger.Cursor.session = property(lambda x: x._session_value)
+def session_open_cursor_notify(cursor, session, *args):
+    cursor._session_value = session
+
+def cursor_notify_for_rollback(cursor, creator):
+    s = cursor.session
+    #tty('   cursor_notify_for_rollback, session={}'.format(s))
+
+    # We only want to trigger a rollback error when we are in a transaction,
+    # to mimic when these errors occur in actual WiredTiger calls.  To do this
+    # we have a field on the session object that tells us if we are in a
+    # transaction.
+    if (hasattr(s, 'in_transaction') and s.in_transaction):
+        if creator.do_retry():
+            # Add something to stdout, we want to make sure it is cleaned up on a test retry.
+            print('stuff in stdout')
```

Comment:
Do we want to print something here that does not look like debugging output? Even if the output is throw-away, maybe something like `print('retry transaction rollback error')`.